runtime/nvqir/cutensornet/tensornet_state.cpp

```diff
@@ -0,0 +1,371 @@
+}
+
+std::unordered_map<std::string, size_t>
+TensorNetState::sample(const std::vector<int32_t> &measuredBitIds,
+                       int32_t shots) {
+  LOG_API_TIME();
+  // Create the quantum circuit sampler
+  cutensornetStateSampler_t sampler;
+  HANDLE_CUTN_ERROR(cutensornetCreateSampler(m_cutnHandle, m_quantumState,
+                                             measuredBitIds.size(),
+                                             measuredBitIds.data(), &sampler));
+
+  ScratchDeviceMem scratchPad;
+  // Configure the quantum circuit sampler
+  constexpr int32_t numHyperSamples =
+      8; // desired number of hyper samples used in the tensor network
+         // contraction path finder
+  HANDLE_CUTN_ERROR(cutensornetSamplerConfigure(
+      m_cutnHandle, sampler, CUTENSORNET_SAMPLER_OPT_NUM_HYPER_SAMPLES,
+      &numHyperSamples, sizeof(numHyperSamples)));
+
+  // Prepare the quantum circuit sampler
+  cutensornetWorkspaceDescriptor_t workDesc;
+  HANDLE_CUTN_ERROR(
+      cutensornetCreateWorkspaceDescriptor(m_cutnHandle, &workDesc));
+  {
+    cudaq::ScopedTrace trace(
+        "TensorNetState::sample::cutensornetSamplerPrepare");
+    HANDLE_CUTN_ERROR(cutensornetSamplerPrepare(m_cutnHandle, sampler,
+                                                scratchPad.scratchSize,
+                                                workDesc, /*cudaStream*/ 0));
+  }
+  // Attach the workspace buffer
+  int64_t worksize{0};
+  HANDLE_CUTN_ERROR(cutensornetWorkspaceGetMemorySize(
+      m_cutnHandle, workDesc, CUTENSORNET_WORKSIZE_PREF_RECOMMENDED,
+      CUTENSORNET_MEMSPACE_DEVICE, CUTENSORNET_WORKSPACE_SCRATCH, &worksize));
+  assert(worksize > 0);
+  if (worksize <= static_cast<int64_t>(scratchPad.scratchSize)) {
+    HANDLE_CUTN_ERROR(cutensornetWorkspaceSetMemory(
+        m_cutnHandle, workDesc, CUTENSORNET_MEMSPACE_DEVICE,
+        CUTENSORNET_WORKSPACE_SCRATCH, scratchPad.d_scratch, worksize));
+  } else {
+    throw std::runtime_error("ERROR: Insufficient workspace size on Device!");
+  }
+
+  // Sample the quantum circuit state
+  std::unordered_map<std::string, size_t> counts;
+  constexpr int MAX_SHOTS_PER_RUNS = 10000;
+  int shotsToRun = shots;
+  while (shotsToRun > 0) {
+    const int numShots = std::min(shotsToRun, MAX_SHOTS_PER_RUNS);
+    std::vector<int64_t> samples(measuredBitIds.size() * numShots);
+    {
+      cudaq::ScopedTrace trace(
+          "TensorNetState::sample::cutensornetSamplerSample");
+      HANDLE_CUTN_ERROR(cutensornetSamplerSample(
+          m_cutnHandle, sampler, numShots, workDesc, samples.data(),
+          /*cudaStream*/ 0));
+    }
+
+    const auto numMeasuredQubits = measuredBitIds.size();
+    for (int i = 0; i < numShots; ++i) {
+      std::string bitstring;
+      constexpr char digits[2] = {'0', '1'};
+      for (std::size_t j = 0; j < measuredBitIds.size(); ++j)
+        bitstring += digits[samples[i * numMeasuredQubits + j]];
+      counts[bitstring] += 1;
+    }
```

Comment:
Consider this optimization to limit the number of allocations in the inner-most loop:
```suggestion
    std::string bitstring(numMeasuredQubits, '0');
    for (int i = 0; i < numShots; ++i) {
      constexpr char digits[2] = {'0', '1'};
      for (std::size_t j = 0; j < numMeasuredQubits; ++j)
        bitstring[j] = digits[samples[i * numMeasuredQubits + j]];
      counts[bitstring] += 1;
    }
```